### PR TITLE
Hesa Winions.

### PR DESCRIPTION
**Pull request by hesa2020**
Progress -- #148

### CODE REVIEW

**Comment by danil179** [comment]
PR seems ok to me. (styleguide ok besides #region tags)
Please open an issue about the problem this PR trying to solve or refer to existing issue,
and add Resolve -- #issue number (see guidelines).

didn't checked if that breaks something.

File: GameServerLib/Geometry/CirclePoly.cs [hunk 1]
```diff
@@ -0,0 +1,74 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Numerics;
+/// <summary>
+///     Represents a Circle <see cref="Polygon" />
+/// </summary>
+public class CirclePoly : Polygon
+{
+    #region Fields
```
Comment (FurkanS1821) [inline GameServerLib/Geometry/CirclePoly.cs]: `#region`s are discouraged by our guidelines.

File: GameServerLib/Logic/Content/NavGrid.cs [hunk 2]
```diff
@@ -652,7 +710,7 @@ private static NavGrid ReadData(NavBinaryReader b)
             grid.MapWidth = grid.MaxGridPos.X + grid.MinGridPos.X;
             grid.MapHeight = grid.MaxGridPos.Z + grid.MinGridPos.Z;
             grid.MiddleOfMap = new Vector2(grid.MapWidth / 2, grid.MapHeight / 2);
-
+            grid.InitializePathfinding();
```
Comment (FurkanS1821) [inline GameServerLib/Logic/Content/NavGrid.cs]: I believe this could be moved to a better function.

File: GameServerLib/Logic/GameObjects/AttackableUnits/AttackableUnit.cs [hunk 3]
```diff
@@ -258,9 +259,9 @@ public enum ClassifyUnit
         TURRET_ATTACKING_MINION = 4,
         CHAMPION_ATTACKING_MINION = 5,
         PLACEABLE = 6,
-        MELEE_MINION = 7,
+        SUPER_OR_CANNON_MINION = 7,
         CASTER_MINION = 8,
-        SUPER_OR_CANNON_MINION = 9,
+        MELEE_MINION = 9,
```
Comment (FurkanS1821) [inline GameServerLib/Logic/GameObjects/AttackableUnits/AttackableUnit.cs]: Whitespace should be removed.

File: GameServerLib/Logic/ObjectManager.cs [hunk 4]
```diff
@@ -380,5 +380,29 @@ public bool TeamHasVisionOn(TeamId team, GameObject o)
...
+        {
+            int count = 0;
+            try
+            {
+                var enemyUnits = GetObjects();
+                foreach (var obj in enemyUnits)
+                {
+                    if(obj.Value is ObjAiBase aiBase)
+                    {
+                        if(aiBase != null && aiBase.Team != target.Team && aiBase.Team != TeamId.TEAM_NEUTRAL && !aiBase.IsDead && aiBase.TargetUnit != null && aiBase.TargetUnit == target)
```
Comment (FurkanS1821) [inline GameServerLib/Logic/ObjectManager.cs]: Lines should not be longer than 120 characters. You might want to split this line (and a few others you have changed/added if they are off limits, too) into multiple lines.

**Review by FurkanS1821 — CHANGES_REQUESTED** [review]
Very good! We've been waiting for a hero to do this for a few years 😄

There are some parts that doesn't comply with our guidelines and I believe this should be tested by multiple users in the meantime.

**Comment by MythicManiac** [comment]
Really nice to see someone working on this 👍 will have to go this over in some detail later this week, but so far from what I gathered in the Discord it sounds like some really good progress 👌

File: GameServerLib/Logic/GameObjects/AttackableUnits/AI/Minion.cs [hunk 5]
```diff
@@ -112,52 +135,65 @@ public override void Update(float diff)
...
             }
-
             Replication.Update();
         }
-
+        
         public override void OnCollision(GameObject collider)
         {
-            if (collider == null || collider == TargetUnit) // If we're colliding with the target, don't do anything.
+            /*if (collider == null || collider == TargetUnit) // If we're colliding with the target, don't do anything.
```
Comment (horato) [inline GameServerLib/Logic/GameObjects/AttackableUnits/AI/Minion.cs:118]: I'd say remove unused blocks of code instead of commeting them out.

File: GameServerLib/Logic/GameObjects/AttackableUnits/AI/Minion.cs [hunk 6]
```diff
@@ -112,52 +135,65 @@ public override void Update(float diff)
...
+            if(TargetUnit != null && !TargetUnit.IsDead)
+            {
+                return true;
+            }
             AttackableUnit nextTarget = null;
             var nextTargetPriority = 14;
 
             var objects = _game.ObjectManager.GetObjects();
-            foreach (var it in objects)
+            //foreach (var it in objects.OrderBy(x => _game.ObjectManager.CountUnitsAttackingUnit(x.Value)))
```
Comment (horato) [inline GameServerLib/Logic/GameObjects/AttackableUnits/AI/Minion.cs:155]: I'd say remove unused blocks of code instead of commeting them out.

File: GameServerLib/Logic/GameObjects/AttackableUnits/AI/ObjAIBase.cs [hunk 7]
```diff
@@ -250,22 +251,26 @@ public void StopMovement()
...
+            if (GetDistanceTo(TargetUnit) <= Stats.Range.Total - 2f)
             {
                 SetWaypoints(new List<Vector2> { new Vector2(X, Y) });
             }
             else
             {
-                var t = new Target(Waypoints[Waypoints.Count - 1]);
-                if (t.GetDistanceTo(TargetUnit) >= 25.0f)
+                SetWaypoints(new List<Vector2>() { GetPosition(), TargetUnit.GetPosition() });
+                /*if(CurWaypoint >= Waypoints.Count)
```
Comment (horato) [inline GameServerLib/Logic/GameObjects/AttackableUnits/AI/ObjAIBase.cs:273]: I'd say remove unused blocks of code instead of commeting them out.

File: GameServerLib/Logic/GameObjects/AttackableUnits/AI/ObjAIBase.cs [hunk 8]
```diff
@@ -555,5 +558,59 @@ public override void OnCollision(GameObject collider)
...
+            List<CirclePoly> UsedPositions = new List<CirclePoly>();
+            var isCurrentlyOverlapping = false;
+
+            var thisCollisionCircle = new CirclePoly((Target != null ? Target.GetPosition() : GetPosition()), CollisionRadius + 10, 20);
+            
+            foreach (var gameObject in objects)
+            {
+                var u = gameObject.Value as AttackableUnit;
+                if (u == null || u.NetId == NetId || u.IsDead || u.Team != Team || u.GetDistanceTo(TargetUnit) > DETECT_RANGE)
+                //if (u == null || u.NetId == NetId || u.IsDead || (u.Team == Team && u.GetDistanceTo(TargetUnit) > DETECT_RANGE))
```
Comment (horato) [inline GameServerLib/Logic/GameObjects/AttackableUnits/AI/ObjAIBase.cs:585]: I'd say remove unused blocks of code instead of commeting them out.

File: GameServerLib/Logic/ObjectManager.cs [hunk 9]
```diff
@@ -380,5 +380,27 @@ public bool TeamHasVisionOn(TeamId team, GameObject o)
...
             return false;
         }
+
+        public int CountUnitsAttackingUnit(AttackableUnit target)
+        {
+            int count = 0;
+            try
+            {
+                var enemyUnits = GetObjects();
+                var team = target.Team == TeamId.TEAM_BLUE ? TeamId.TEAM_PURPLE : TeamId.TEAM_BLUE;
```
Comment (horato) [inline GameServerLib/Logic/ObjectManager.cs:390]: You can use GetEnemyTeam(TeamId) extension method from CustomConvert
https://github.com/LeagueSandbox/GameServer/blob/828e0f7714a644ed2aff87a0364a57da5800efe0/GameServerCore/Extensions.cs#L181

File: GameServerLib/Logic/ObjectManager.cs [hunk 10]
```diff
@@ -380,5 +380,27 @@ public bool TeamHasVisionOn(TeamId team, GameObject o)
...
+                count = enemyUnits.Count(x =>
+                    x.Value is ObjAiBase aiBase &&
+                    aiBase.Team == team &&
+                    aiBase.IsDead == false &&
```
Comment (horato) [inline GameServerLib/Logic/ObjectManager.cs]: not sure how I feel about comparing boolean variables to true/false. Why not just use 
`!aiBase.IsDead`

File: GameServerLib/Logic/Game.cs [hunk 11]
```diff
@@ -129,42 +131,42 @@ public bool LoadScripts()
...
+                
                 if (IsPaused)
                 {
                     _lastMapDurationWatch.Stop();
                     _pauseTimer.Enabled = true;
                     if (PauseTimeLeft <= 0 && !_autoResumeCheck)
                     {
-                        PacketHandlerManager.GetHandler(PacketCmd.PKT_UNPAUSE_GAME, Channel.CHL_C2S)
-                            .HandlePacket(null, new byte[0]);
+                        PacketHandlerManager.GetHandler(PacketCmd.PKT_UNPAUSE_GAME, Channel.CHL_C2S).HandlePacket(null, new byte[0]);
```
Comment (horato) [inline GameServerLib/Logic/Game.cs]: Not sure why this change is here since nothing seems to be changed, but it is not gonna work, use IPacketHandlerManager.UnpauseGame()
https://github.com/LeagueSandbox/GameServer/blob/828e0f7714a644ed2aff87a0364a57da5800efe0/GameServerCore/Packets/Handlers/IPacketHandlerManager.cs#L21

**Comment by horato** [comment]
You should rebase with indev first, there are some conflicts.

**Comment by MythicManiac** [comment]
Builds also seem to fail

**Comment by MythicManiac** [comment]
I also feel like if we want to use an external library, we shouldn't embed the actual source code into the project. At least make it it's own project solution wise, or better yet, use a Nuget package.

**Comment by hesa2020** [comment]
Are you talking about ClipperLib?
There is no package and this file was edited, i can make it a lib tho.
How do you want embed it.

**Comment by MythicManiac** [comment]
@hesa2020 Just a separate project in the solution should suffice for now
